### DISCUSSION TITLE
External sensors

### DISCUSSION THREAD

**Discussion by lgordon** [martin2250/OpenCNCPilot#180] [Ideas]
I'm interested in building a low cost optical profilometer based on height/distance sensors and CNC hardware. I was wondering what the recommended way would be to modify/extend openCNCpilot to record the height to the workpiece from an external sensor mounted in the tool holder/spindle instead of using z-probing? Does this seem like something that might be somewhat straightforward to do?

**Comment by deHarro** [comment]
Hi!
Just to get it right...
Your idea is to _not_ do touch probing as it is realized in OCP at the moment, but you think of some sort of scanning the hight map while meandering over the to be scanned surface without Z moves.
Correct?
If so, this would be very fast compared to the implemented scenario.

In this case OCP has to emit an impulse for every dedicated coordinate where a measure should be taken to trigger the measure of your sesnsor, record the coordinate and read the distance from your sensor.

The problem I see: There is no interface in OCP other than to GRBL via serial communication over USB.

Ok, let Martin give his thoughts about it :)

Harald

**Comment by martin2250** [comment]
Hi Lyle,

Harald is right, GRBL has no way to interface such sensors so you'd have to implement the interface in OCP. Also I don't know of any "standard" sensors from the 3D printing community. Since I have neither time nor the hardware, I'm not going to implement that for the standard version of OCP. That said, it should be pretty straightforward to do if you know your way around C#.

If you have any questions regarding how to do that, feel free to ask.

Cheers, Martin
